### ISSUE TITLE
Unexpected type error when extending similar interfaces with optional fields

### ISSUE THREAD

**Issue by AviVahl** [Microsoft/TypeScript#27489]
**TypeScript Version:**  3.1.1 & 3.2.0-dev.20180929

**Search Terms:**
`cannot simultaneously extend types`

**Code**
(you might appreciate this example, as the actual use case was using the typescript programmatic API)
```ts
import * as ts from 'typescript'
interface BaseHost extends ts.ParseConfigHost, ts.FormatDiagnosticsHost, ts.ModuleResolutionHost {}
```

**Expected behavior:**
No error.

**Actual behavior:**
```
[ts]
Interface 'BaseHost' cannot simultaneously extend types 'FormatDiagnosticsHost' and 'ModuleResolutionHost'.
  Named property 'getCurrentDirectory' of types 'FormatDiagnosticsHost' and 'ModuleResolutionHost' are not identical.
```
That's because getCurrentDirectory is optional on `ModuleResolutionHost`, but not the other hosts.

Is there a reason to not make the field in question **required** on `BaseHost` instead of throwing an error?
Type-wise, would this not yield correct behavior? Meaning any `BaseHost` is indeed compatible as the three TypeScript interfaces it extends.

**Playground Link:** 
http://www.typescriptlang.org/play/#src=interface%20A%20%7B%0D%0A%20%20%20%20someField%3A%20string%0D%0A%7D%0D%0A%0D%0Ainterface%20B%20%7B%0D%0A%20%20%20%20someField%3F%3A%20string%0D%0A%7D%0D%0A%0D%0Ainterface%20C%20extends%20A%2C%20B%20%7B%0D%0A%0D%0A%7D%0D%0A

Playground example is a much simpler use case.

**Related Issues:** 
https://github.com/Microsoft/TypeScript/issues/16936 - one could say this is a subset of that suggestion. the scope here is just optional fields

**Comment by AviVahl** [comment]
I should also note that using:
```ts
type BaseHost = ts.ParseConfigHost & ts.FormatDiagnosticsHost & ts.ModuleResolutionHost
```
simply destroyed any type inference for object method parameters:
```
function someFunc(): BaseHost {
    return {
        ...,
        readFile(path) {}
    }
}
```
(inferred type of path is lost. no longer `string` by default)

**Comment by RyanCavanaugh** [comment]
Going from the example
```ts
interface A {
    someField: string
}

interface B {
    someField?: string
}

interface C extends A, B {

}
```
Is this code legal?
```
const c: C = { someField: "" };
c.someField = undefined;
```
 * Argument 1: It has to be, because that's a legal operation on a `B` and `C` claims to be substitutable for `B`
 * Argument 2: It can't be, because `A` claims to always have a non-undefined `someField`, and `C` violates that

There's not a principled way to choose one over the other, so we defer the question to the `C` author to resolve the dispute.

**Comment by AviVahl** [comment]
Never mind, I realized I can simply:
```ts
interface IBaseHost extends ts.ParseConfigHost, ts.FormatDiagnosticsHost, ts.ModuleResolutionHost {
    getCurrentDirectory(): string
}
```
and I also guess that's what you meant by saying the decision is deferred to me. :+1: 

Taking your original explanation into consideration, I'm no longer sure how this is any more consistent. :) Because the following now passes type checking, and the author of run might not be
```ts
declare const aaa: BaseHost
function run(host: ts.ModuleResolutionHost) {
    host.getCurrentDirectory = undefined
}
run(aaa)
```

**Comment by AviVahl** [comment]
Closing, as my original issue is mostly solved. Although I ended up seeing more holes in the system :)

**Comment by AviVahl** [comment]
Thanks for the kind explanation again. I guess you guys get these kind of tickets a lot from people like me not always seeing the delicate nature between usability and safety of the type system. The use case of `run()` made me question myself again, thinking this might be a usability intended decision.

**Comment by RyanCavanaugh** [comment]
> Never mind, I realized I can simply:

I really should have included that in my example!
